[TensorExpr] Start using typecheck in the fuser.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43603 Benchmarks: temporarily disable profiling-te configuration.
* **#43173 [TensorExpr] Start using typecheck in the fuser.**
* #43235 [JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings.
* #43365 [TensorExpr] Fuser: disallow aten::slice nodes.
* #43171 [TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred).
* #43170 [TensorExpr] Fuser: move canHandle and some other auxiliary functions into TensorExprFuser class.

With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

Commit:
[WIP][TensorExpr] Fuser changes.

Not ready for commit.

[ghstack-poisoned]

Commit:
Update on "[WIP][TensorExpr] Fuser changes."


Not ready for commit.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      // TODO: Make sure we use types encoded in the IR (not through the
+      // typeinfo map, but explicitly as value types)
+      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
```
Could you add a comment somewhere about how we only check the input types and use subgraph intermediary types for free because currently NNC only handles ops whose output types are uniquely determined by their inputs

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -177,7 +187,9 @@ class TensorExprFuser {
     if (minSubgraphSize_ == 1) {
-      fusion_group = getOrCreateTensorExprSubgraph(n);
+      std::unordered_map<Value*, Value*> vmap;
+      fusion_group = getOrCreateTensorExprSubgraph(n, vmap);
```
nitpick: Should we consider making `getOrCreateTensorExprSubgraph` a member method of `TensorExprFuser` so we don't have to thread `vmap`?

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -196,6 +208,34 @@ class TensorExprFuser {
       updateQueue(fusion_group, queue, visited_nodes);
     }
 
+    // Update typeinfo for inputs and outputs of the fusion group
```
nitpick: can we refactor this into a helper as this blob of logic has a very specific purpose.

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
+          GRAPH_DEBUG("TYPEINFO ERASE: %", o->debugName());
+          typeinfo_map_.erase(o);
+        }
+        if (it->outputs().size()) {
```
we should be able to remove this check right ? or add a TODO for the rebase

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
 
+  void fillTypeInfoMap(Block* block) {
+    for (Node* n : block->nodes()) {
+      if (n->kind() == prim::profile && n->outputs().size() == 1) {
```
nit; should be able to remove this check on the rebase

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
+    }
+  }
+
+  void removeProfilingNodes(Block* b) {
```
nitpick: I'm not sure `removeProfilingNodes` belongs in this file. `profiling_record.h`, `profiling_graph_executor_impl.h` might be a better choice.

ZolotukhinM replied:
Yeah, it probably doesn't belong here. I plan to do a spring cleaning of PE-related passes after our changes land and stabilize a little, and as a part of it I'll move this function to a better place.

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
+    }
+  }
+
+  std::unordered_map<Value*, TensorTypePtr> typeinfo_map_;
```
maybe add a comment somewhere that the specialized tensors will be Value * in the graph before they are merged, and Value * in the fusion subgraph after they are merged

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -196,6 +208,34 @@ class TensorExprFuser {
       updateQueue(fusion_group, queue, visited_nodes);
     }
 
+    // Update typeinfo for inputs and outputs of the fusion group
```
So, the reason we're not updating the typeinfo map until after we've already exhausted the queue is because we never expect to successfully merge two Fusion groups together. I think this is always the case but I'm not totally sure. It might be worth updating the fusion group outputs as part of updateTypeinfoMapWithVmap instead of deferring it till after

eellison replied:
based on conversation today, I guess we do expect to merge two fusion groups together

ZolotukhinM replied:
That is correct observation, I'm going to revisit this part as I'll be working on merging adjacent fusion groups.

eellison reviewed (commented):
This looks good!! I have only minor comments & nits.

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -143,7 +145,15 @@ class TensorExprFuser {
 
   void run() {
     aliasDb_ = torch::make_unique<AliasDb>(graph_);
+    fillTypeInfoMap(graph_->block());
```
Yeah, good idea, I'll do that in a follow up if you don't mind!

Krovatkin replied:
yeah let's do it in a separate PR

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
+    // Fixup types of the subgraph inputs
+    std::vector<Value*> inputs_to_check;
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      // TODO: Make sure we use types encoded in the IR (not through the
```
Reworded!

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Commit:
flake8 on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -260,14 +307,26 @@ class TensorExprFuser {
   }
 
+  void updateTypeinfoMapWithVmap(
+      const std::unordered_map<Value*, Value*>& vmap) {
```
i'd love to see an assertion here to double check that we don't have the same address as both a key and value. Took two hours out of my life. But we could do it later.

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -295,17 +354,23 @@ class TensorExprFuser {
     for (auto n : nodes_to_merge) {
       GRAPH_UPDATE("Merging ", getHeader(n));
-      SubgraphUtils::mergeNodeIntoSubgraph(n, fusion_group);
+      SubgraphUtils::mergeNodeIntoSubgraph(n, fusion_group, vmap);
```
can we wrap mergeNodeIntoSubgraph and updateTypeinfoMapWithVmap into one single helper? Otherwise, there's always a possibility somewhat will insert another graph transformation and now we are running into the same issue of having the same address as a key and a value.

eellison replied:
What does this mean?
> there's always a possibility somewhat will insert another graph transformation

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -396,18 +458,168 @@ class TensorExprFuser {
+        typecheck_node->output(output_idx++)
+            ->setType(typeinfo_map_.at(subgraph->inputs()[idx]));
+      }
+    }
```
can we consider adding an assertion here in the else clause that `subgraph->inputs()[idx]->isCompleteTensor()` if it's of a tensor type?

Krovatkin reviewed (changes requested):
:shipit:

eellison reviewed (approved):
🚢

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Commit:
clang-format on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Start using typecheck in the fuser."


With this change the fuser starts to generate typechecks for inputs of
fusion group. For each fusion group we generate a typecheck and an if
node: the true block contains the fused subgraph, the false block
contains unoptimized original subgraph.

Differential Revision: [D23178230](https://our.internmc.facebook.com/intern/diff/D23178230)

[ghstack-poisoned]
